frontend for webapp

Dami-18 opened the pull request:
Fixes #11 
Frontend for the webapp created using React typescript

Dami-18 commented:
This is the basic layout. CSS still needs to be added and also login form is not yet complete.

Dami-18 commented:
I have added basic design. It is minimal. In the empty space, graphics can be added
Also, yet to add media queries. Will do it once finalized.

rohan-b-84 commented on frontend/README.md:
Do update the README before merging

Dami-18 replied:
Ok
And I didn't get about the UI. Like I am thinking to add some graphics in the empty space. Some graphics matching with acads and colorful so that they match with the background.

rohan-b-84 replied:
For UI, you'll have to consider the possibility that we will show the breadth/depth in the UI itself in future, so keep that in consideration

rohan-b-84 commented on frontend/src/App.tsx:
```diff
@@ -0,0 +1,28 @@
+        <Route path="/" element={<Form/>}/>
+        <Route path="/login" element={<Login/>}/>
+        <Route path="/electives" element={<Electives/>}/>
+      </Routes>
```
you don't need to different routes for each step. Try multi-step form without the need of route.
Also electives should be a protected route only available after login, right now I can go to browser and type /electives and it'll work

Dami-18 replied:
Ok have never done this. Will look into it

rohan-b-84 replied:
let me know if you need anything

rohan-b-84 commented on frontend/src/assets/metakgp-logo.png:
remind me to provide up-scaled metakgp logo tomorrow

Dami-18 replied:
Can you please provide scaled-up logo?

rohan-b-84 commented on frontend/src/components/Electives.tsx:
Looks good, maybe add placeholder handlers for the buttons, and some styling.

Dami-18 replied:
Yes styling is left. Will add that

rohan-b-84 commented on frontend/src/components/Login.tsx:
The login process is a three step process:
1. Once you enter roll number -> we need to request for security question
2. Then send roll number, password, and the answer to security question to request otp via email
3. Send roll number, password, security question's answer and the received otp -> get logged in

We'll need to preserve states between these requests, so I'll suggest proper state handing for user's input.

Dami-18 replied:
ok

rohan-b-84 reviewed (changes requested):
This looks good! 
Just add some more structure to it in terms of state handling. This way, once the API is finished, there won't be much leftover work.

Try to make your components responsive from the start. Also, do you have any plans for the UI ? we can discuss the UI enhancement next once you're done with these changes.

Dami-18 commented:
I have tried to create protected route but still back navigation causes problem. How to solve it? Disable back navigation after login? 
And to handle user inputs, I used getValues from useForm hook. These values will be then sent to api where they will be saved. Is it correct? What further changes? 
I will add placeholder handlers for buttons later.

rohan-b-84 commented on frontend/src/App.tsx:
```diff
@@ -1,23 +1,27 @@
-        <Route path="/" element={<Form/>}/>
-        <Route path="/login" element={<Login/>}/>
-        <Route path="/electives" element={<Electives/>}/>
+        <Route path="/" element={<MultiForm updateStatus={updateStatus}/>}/>
```
1. Download electives page should only be visible when authenticated. Likewise, login page should only be visible when not logged in.
2. Consider renaming the routes to `/login` (for auth) and `/` (for electives)
3. if not authenticated, can redirect to `/login`, else the electives page. The error page can be used as catch-all for all errors

Dami-18 replied:
Solved the issue. I guess routing works fine now.

rohan-b-84 commented on frontend/src/App.tsx:
```diff
@@ -1,23 +1,27 @@
 
 const App: React.FC = () => {
 
+  const [isLoggedIn, setLogin] = useState(false);
```
Preserve the state on reload here

rohan-b-84 reviewed (changes requested):
Also, I think we can include an about/help section in this.

rohan-b-84 commented on frontend/src/App.tsx:
```diff
@@ -1,15 +1,21 @@
+  });
+
+  useEffect(() => {
+    sessionStorage.setItem('loginStatus',isLoggedIn.toString());
```
later it would store the JWT. but I think its fine for now.

rohan-b-84 commented on frontend/src/components/Header.tsx:
```diff
@@ -1,5 +1,5 @@
 import React from 'react';
-import metakgpLogo from '../assets/metakgp-logo.png';
+import metakgpLogo from '../assets/header-logo3.png';
```
keep names consistent, if `header-logo.png` not being used, you can remove the older file, and rename correct one to `header-logo.png`

Dami-18 commented:
Using session storage, we have access to roll number and password entered in previous form.

Dami-18 commented:
fixes #11

Dami-18 commented:
fixes #11

Dami-18 commented:
Do I need to keep an option for explicit login?

Dami-18 commented:
Integration of api with electives page is still left

Dami-18 commented:
Please review this. I have added global state management and completed integration with api. (From latest changes in Prem's PR)

Dami-18 commented:
I have changed:

1. Decreased the spacing between logo and header
2. Changed the the layout of download electives page as mentioned
3. Switched to dark mode

Dami-18 commented:
@rohan-b-84 I tried to change the layout to make that input boxes fixed but messed up with responsiveness. So reverted it. Also the two divs in electives page are not responsive in mobile and I am unable to debug it. Can you please work on this if possible?
Currently I am trying that  user input feature for electives, logout option and a work around for sessionStorage of login status

Dami-18 commented:
Made electives page responsive.

Dami-18 commented:
I tried global state but the problem is that on reload, all are set to default values which are empty and user data is lost. So logged in state is not saved. So, I think we have only two options for this, to use cookies or use sessionStorage. What should be done now?

Dami-18 commented:
I figured out alignment of boxes and also not showing login button until security question. So, I am working on it

Dami-18 commented:
Fixed the UI

Dami-18 commented:
Need to do some changes in Electives.tsx. I am doing that now

rohan-b-84 reviewed (approved):
Approved, great work @Dami-18  !!
